### PR TITLE
Refactoring XmlConvert

### PR DESCRIPTION
**Pull request by kronic**

### CODE REVIEW

**Review by eiriktsarpalis — COMMENTED** [review]
Hi @kronic, thanks for submitting this PR. Superficially the changes look good to me, however please note that they might not meet the bar according to our [contribution guidelines](https://github.com/dotnet/runtime/blob/main/CONTRIBUTING.md#contributing-changes), more specifically the following points:

> * **DON'T** make PRs for style changes.
> * **DON'T** surprise us with big pull requests. Instead, file an issue and start a discussion so we can agree on a direction before you invest a large amount of time.

In other words, we might not accept "refactoring" PRs unless they are meant to address a specific set of concerns (e.g. performance improvements), but these should be clearly communicated to the maintainers in either a preceding issue or in the PR description. Every change carries the risk of accidental regressions, so from our perspective it is important to understand the motivations behind every PR that we review.

We might still consider style changes, but these would typically need to be targeted changes that affect the style of the entire project (e.g. "Use the `??=` operator in all of System.Xml" or "Use `ArgumentNullException.ThrowIfNull` method instead of throwing in all of System.Xml" or "Use target typed constructors in System.Xml") but these would need to be merged as separate commits.

This feedback also applies to some of the other PRs I've seen, e.g. #60074.

I defer to @krwq on a final review of this PR.

File: src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs:429]: Cool. In general I would say that the coding conventions local to the current project take precedence over our current guidelines. So if the naming of `crt` specifically is an outlier then it's fine to change; on the other hand if missing `s_` prefixes are pervasive then I would prefer it if they were all changed in one dedicated PR.

Comment (kronic) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs:429]: @Eiriktsarpalis Undo this in this PR?

Comment (eiriktsarpalis) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs:429]: Up to you. The change seems consistent for the particular source file, but I wouldn't want it be inconsistent with other source files in System.Xml

File: src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs
Comment (krwq) [inline src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs:1219]: consider leaving the comment

File: src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs
Comment (krwq) [inline src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs:1206]: this is potentially slower because CreateAllDateTimeFormats is not guaranteed to be inlined.

File: src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs
Comment (krwq) [inline src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs:1190]: split line enum flags look harder to read IMO

File: src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs
Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs:1048]: I also find more condensed style easier to read

File: src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs
Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs:835]: I personally find it easier to read as a single line

File: src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs
Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs:429]: `s_crt` looks good to me

**Review by krwq — APPROVED** [review]
Overall looks good, please revert some of the changes (i.e. splitting enum flags into multi-line and everywhere where there is doubt that something looks cleaner) - remainder looks good and thank you for the cleanup

**Comment by kronic** [comment]
@krwq done

File: src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs
Comment (krwq) [inline src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs:199]: is this actually nullable? I'd expect you get warning on the next line, given that you haven't I assume it's non-null

File: src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs
Comment (krwq) [inline src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs:212]: Please leave the extra brackets here - It's not as obvious what order those are executed

File: src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs
Comment (krwq) [inline src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs:245]: same about the brackets, I'm not really sure this is equivalent...

Comment (krwq) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs:245]: (and anywhere further where applies)

File: src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs
Comment (krwq) [inline src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs:311]: @stephentoub mentioned to undo this because of the future `!!` operator

**Comment by kronic** [comment]
@krwq done

**Comment by jeffhandley** [comment]
I've marked this PR as "needs more info" for now. We'd like to better understand the risk/reward for the series of XML changes. See https://github.com/dotnet/runtime/pull/61773#issuecomment-973646209. Thanks!

**Comment by kronic** [comment]
I gave my answer [Comment](https://github.com/dotnet/runtime/pull/61773#issuecomment-973936099)

**Comment by krwq** [comment]
@kronic please create an uber issue with motivation for refactoring and (more or less) planned scope of changes before we proceed with merging this. I'm ok with proceeding with PRs which got sent before we've said anything but still issue we could link that to would be useful. Before sending any future refactoring PRs I'd like to see a motivation and some discussion around it first.
I'm personally seeing value in optimization in general (especially where it gives measurable improvements in end to end scenarios) but for refactoring I got mixed feelings: it makes code simpler and easier to read for the future but at the same time it doesn't have any value from consumer point of view, takes a bit of time to review and there is some risk of regression - therefore for pure refactoring I'd prefer we've made some buckets on what types of changes are we planning so both reviewing is easier and risk of regression is smaller

**Pull request closed without merging**
